File: tmp/leetcode75/57.py
'''
heaps:
less = max heap
greater = min heap

addnum:
if len(less) > len(greater):
    if num <= median (less)
        add to less
        remove from less, add to greater
    if num > median (less)
        add to greater
if len(greater) > len(less):
    if num <= median (greater)
        add to less
    if num > median (greater)
        add to greater
        remove from greater, add to less
elif len(less) == len(greater):


findmedian:
if len(less) == len(greater):
    average of top of both
elif len(less) > len(greater):
    top of less
else
    top of greater
'''
import heapq

class MedianFinder:

    def __init__(self):
        self.less = [] # max heap
        self.greater = [] # min heap

    def addNum(self, num: int) -> None:
        heapq.heappush(self.less, -num)
        heapq.heappush(self.greater, -heapq.heappop(self.less))
        # Ensure that less is >= greater by 1
        if len(self.less) < len(self.greater):
            heapq.heappush(self.less, -heapq.heappop(self.greater))

    def findMedian(self) -> float:
        if len(self.less) == len(self.greater):
            return (-self.less[0] + self.greater[0])/2
        elif len(self.less) > len(self.greater):
            return -self.less[0]

# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: tmp/leetcode75-c/57.c
void swap(int *x, int *y) {
    int tmp = *x;
    *x = *y;
    *y = tmp;
}

void push(int **heap, int *n, int *sz, int val) {
    if (*n == *sz) {
        *sz *= 2;
        *heap = realloc(*heap, *sz * sizeof **heap);
    }
    (*heap)[*n] = val;
    *n += 1;

    int i = *n-1, parent = (i-1)/2;
    while (i != 0 && (*heap)[parent] > (*heap)[i]) {
        swap(&(*heap)[parent], &(*heap)[i]);
        i = parent;
        parent = (i-1)/2;
    }
}

void cascade(int *heap, int n, int i) {
    int child, min = i;
    for (;;) {
        child = i << 1;
        if (++child < n && heap[child] < heap[min])
            min = child;
        if (++child < n && heap[child] < heap[min])
            min = child;

        if (min == i)
            return;

        swap(&heap[min], &heap[i]);
        i = min;
    }
}

int pop(int *heap, int *n) {
    int min = heap[0];
    heap[0] = heap[*n-1];
    *n -= 1;
    cascade(heap, *n, 0);
    return min;
}


typedef struct {
    int *less;
    int less_n;
    int less_sz;
    int *greater;
    int greater_n;
    int greater_sz;
} MedianFinder;

/** initialize your data structure here. */

MedianFinder* medianFinderCreate() {
    MedianFinder *mf = malloc(sizeof(*mf));
    mf->less_n = 0;
    mf->less_sz = 10;
    mf->less = malloc(mf->less_sz * sizeof(*mf->less));
    mf->greater_n = 0;
    mf->greater_sz = 10;
    mf->greater = malloc(mf->greater_sz * sizeof(*mf->greater));
    return mf;
}

void medianFinderAddNum(MedianFinder* obj, int num) {
    push(&obj->less, &obj->less_n, &obj->less_sz, -num);
    push(&obj->greater, &obj->greater_n, &obj->greater_sz, -pop(obj->less, &obj->less_n));
    if (obj->less_n < obj->greater_n)
        push(&obj->less, &obj->less_n, &obj->less_sz, -pop(obj->greater, &obj->greater_n));
}

double medianFinderFindMedian(MedianFinder* obj) {
    if (obj->less_n == obj->greater_n)
        return ((double) -obj->less[0] + (double) obj->greater[0])/2.0;
    else if (obj->less_n > obj->greater_n)
        return -obj->less[0];
    return 0;
}

void medianFinderFree(MedianFinder* obj) {
    free(obj->less);
    free(obj->greater);
    free(obj);
}

/**
 * Your MedianFinder struct will be instantiated and called as such:
 * MedianFinder* obj = medianFinderCreate();
 * medianFinderAddNum(obj, num);

 * double param_2 = medianFinderFindMedian(obj);

 * medianFinderFree(obj);
*/
